### PR TITLE
Feature/resettle

### PR DESCRIPTION
**Pull request by shanefontaine**
Some minor architecture changes and some bug fixes related to re-settling transfer roots.

This covers all cases, I believe. Some cases that were tested:
- All transfers bonded -> committed
- Most transfers bonded, one withdrew -> committed
- Most transfers bonded -> committed -> settled -> last transfer bonded
- Most transfers bonded -> committed -> settled -> last transfer withdrawn
- Most transfers bonded -> committed -> settled -> transfer bonded -> settled -> last transfer bonded
- Similar flows as above with multiple bonders
- Initial sync (parallel event processing)

### CODE REVIEW

**Review by miguelmota — APPROVED** [review]

**Comment by miguelmota** [comment]
I think the changes make a lot of sense 

Left some minor comments

File: packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts [hunk 1]
```diff
@@ -86,52 +90,29 @@ class SettleBondedWithdrawalWatcher extends BaseWatcher {
...
           continue
         }
-        bonderSet.add(dbTransfer.withdrawalBonder)
+        bonderSet.add(dbTransfer.withdrawalBonder!) // eslint-disable-line @typescript-eslint/no-non-null-assertion
       }
 
       for (const bonder of bonderSet.values()) {
-        const bonderAddress = await this.bridge.getBonderAddress()
-        if (bonder !== bonderAddress) {
-          continue
```
Comment (miguelmota) [inline packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts:116]: is it expected that a bonder may settle another bonders unsettled transfers?

File: packages/hop-node/src/watchers/SyncWatcher.ts [hunk 2]
```diff
@@ -857,21 +856,18 @@ class SyncWatcher extends BaseWatcher {
...
     const tree = new MerkleTree(_transferIds)
     const computedTransferRootHash = tree.getHexRoot()
     if (computedTransferRootHash !== transferRootHash) {
       logger.warn(
         `populateTransferRootTimestamp computed transfer root hash doesn't match. Expected ${transferRootHash}, got ${computedTransferRootHash}. isNotFound: true, List: ${JSON.stringify(_transferIds)}`
       )
       await this.db.transferRoots.update(transferRootHash, { isNotFound: true })
-    } else {
-      await this.db.transferRoots.update(transferRootHash, {
-        transferIds: _transferIds
```
Comment (miguelmota) [inline packages/hop-node/src/watchers/SyncWatcher.ts:871]: is it expected that the transferRoot item should always have the list of transferIds at this point if we're removing the update here?

Comment (shanefontaine) [reply packages/hop-node/src/watchers/SyncWatcher.ts:871]: Great catch. For some reason I thought that the db `transferIds` was compared to the calldata `_transferIds`, so they would equal here. I see that is a bad assumption. Fixed with 442503ad1eb91248e130323f1b4ded9df32790c5

File: packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts [hunk 3]
```diff
@@ -42,30 +43,33 @@ class SettleBondedWithdrawalWatcher extends BaseWatcher {
...
-        }
+      // Mark a settlement as attempted here so that multiple db reads are not attempted every poll
+      // This comes into play when a transfer is bonded after others in the same root have been settled
+      if (!this.settleAttemptedAt[transferRootHash]) {
+        this.settleAttemptedAt[transferRootHash] = 0
+      }
+      const timestampOk = this.settleAttemptedAt[transferRootHash] + OneHourMs < Date.now()
+      if (!timestampOk) {
+        return
```
[the remark this answers is not in the document]
Comment (shanefontaine) [reply packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts]: Yes. Fixed with 442503ad1eb91248e130323f1b4ded9df32790c5

File: packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts (hunk 1 above)
Comment (shanefontaine) [reply packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts:116]: At this time, yes. This comes into play when transactions are manually bonded.

In an ideal world, the bonder themself bonds everything and only settles their own transfers, but given some manual bonds and potential edge-case bonder swaps, I believe it makes sense to bond all at this time. Let me know if you agree with this.

Comment (miguelmota) [reply packages/hop-node/src/watchers/SettleBondedWithdrawalWatcher.ts:116]: I'm good with that

**Pull request merged**